sale_product_country_restriction/models/sale_order.py

```diff
@@ -0,0 +1,109 @@
+        string="Country Restriction",
+        readonly=True,
+    )
+
+    @api.model
+    def _get_no_country_restriction_partner_message(self, partner):
+        return _("The country of the partner %s must be set") % partner.display_name
+
+    @api.model
+    def _get_no_restriction_partner_message(self, partner):
+        self.ensure_one()
+        return (
+            _("A country restriction of the partner %s must be set")
+            % partner.display_name
+        )
+
+    def _check_partner_shipping_country_restriction(self):
+        """
+        Country Restriction on Partner is mandatory
+        :return:
+        """
+        for partner in self.mapped("partner_shipping_id"):
+            if partner and not partner.country_restriction_id:
+                raise ValidationError(
+                    self._get_no_country_restriction_partner_message(partner)
+                )
+            if partner and not partner.country_id:
+                raise ValidationError(self._get_no_restriction_partner_message(partner))
+
+    def _get_country_restriction_products_to_check(self):
+        """
+        This is a hook if one wants to filter checked products
+        :return:
+        """
+        self.ensure_one()
+        return self.order_line.mapped("product_id")
+
+    def check_country_restriction(self):
+        restriction_obj = self.env["product.country.restriction"]
+        self._check_partner_shipping_country_restriction()
+        for order in self:
+            countries = order.partner_shipping_id.country_id
+            products = order._get_country_restriction_products_to_check()
+            restrictions = products._get_country_restrictions(
+                countries, restriction_id=order.country_restriction_id
+            )
+            if restrictions:
+                messages = restriction_obj._get_country_restriction_messages(
+                    restrictions
+                )
+                if messages:
+                    raise ValidationError(messages)
+
+    def action_confirm(self):
+        if self.env.company.enable_sale_country_restriction:
+            self.check_country_restriction()
+        return super().action_confirm()
+
+    @api.onchange("partner_shipping_id")
+    def _onchange_partners_check_country(self):
+        res = {}
+        company = self.company_id or self.env.company
+        if company.enable_sale_country_restriction:
+            for partner in self.mapped("partner_shipping_id"):
+                if partner and not partner.country_id:
+                    warning = (
+                        _("The country of the partner %s must be set")
+                        % self.partner_shipping_id.display_name
+                    )
+                    res = _append_warning(res, warning)
+        return res
+
+    @api.onchange("partner_shipping_id")
+    def _onchange_partners_check_restriction(self):
+        res = {}
+        if not self.env.company.enable_sale_country_restriction:
```

Comment:
Same here: you should use the company of the SO first (as it's `onchange`, it could be empty).